app/controllers/persons_controller.rb

```diff
@@ -18,84 +14,66 @@ def index
+      .flatten
+      .sort
+      .uniq
+  end
 
-  def return_people(all, persons)
-    people = []
-    if params.has_key?("specialists")
-      people = persons
-    end
-
-    if params.has_key?("specialty")
-      if people.blank?
-        special_people = specialties_list(persons).select { |p| p.id }
-      else
-        special_people = specialties_list(people).select { |p| p.id }
-      end
-    end
-    if params.has_key?("department")
-      if people.blank?
-        department_people = departments_list(persons).select { |p| p.id }
-      else
-        department_people = departments_list(people).select { |p| p.id }
-      end
-    end
-    if params.has_key?("location")
-      if people.blank?
-        location_people = locations_list(persons).select { |p| p.id }
-      else
-        location_people = locations_list(people).select { |p| p.id }
-      end
-    end
+  def get_location_filter_values(people)
+    people
+      .map(&:spaces)
+      .flatten
+      .sort
+      .uniq
+  end
 
-    filtered_people = [Array(location_people), Array(department_people), Array(special_people)].reject(&:empty?).reduce(:&) || []
+  def get_persons
+    @filtered_persons = Person
+                        .is_specialist(params[:specialists])
+                        .with_specialty(params[:specialty])
+                        .at_location(params[:location])
+                        .in_department(params[:department])
+                        .order(:last_name, :first_name)
+
+    @persons_list = @filtered_persons.page params[:page]
+    return_filters(@filtered_persons)
+  end
 
-    @people = filtered_people.presence || persons
-    @people_list = Person.where(id: @people.map(&:id)).order(:last_name, :first_name)
-    get_filters(@people_list)
-    @persons_list = @people_list.page params[:page]
+  def return_filters(filtered_persons)
+    @subjects = get_specialty_filter_values(filtered_persons)
+    @departments = get_department_filter_values(filtered_persons)
+    @locations = get_location_filter_values(filtered_persons)
   end
 
-  def get_filters(people)
-    @subjects = @people.select { |person| person.specialties.try(:any?) }.collect { |p| p.specialties }.flatten.uniq.sort.reject { |s| s.empty? }
-    @departments = @people.select { |person| person.groups.try(:any?) }.collect { |p| p.groups }.flatten.uniq.sort.reject { |g| g.group_type != "Department" }
-    @locations = @people.select { |person| person.spaces.try(:any?) }.collect { |p| p.spaces }.flatten.uniq.sort
+  def department
+    params[:department]
   end
 
+  def location
+    params[:location]
+  end
 
   private
     def set_person
       @person = Person.friendly.find(params[:id])
-      @department = @person.groups.collect.reject { |g| g.group_type != "Department" }.uniq.sort
+      @department = @person.groups.collect { |g| g.group_type == "Department" }.uniq.sort
       @location = @person.spaces.collect.uniq.sort
     end
```

Comment:
Break out the processing steps into separate lines for department and location as noted above.

Reply:
Removed lines as redundant code.